dashboard/app/assets/javascripts/header.js

```diff
@@ -73,6 +73,20 @@ dashboard.buildHeader = function (stageData, progressData, currentLevelId, userI
     }
     progressContainer.append(div).append('\n');
   });
+
+  $.ajax('/api/user_progress/' + scriptName + '/' + stageIndex).done(function (data) {
+    // Merge progress from server (loaded via AJAX)
+    var serverProgress = data || {};
+    Object.keys(serverProgress).forEach(function (levelId) {
+      if (serverProgress[levelId] !== clientProgress[levelId]) {
+        var status = mergedActivityCssClass(clientProgress[levelId], serverProgress[levelId]);
+
+        // Clear the existing class and replace
+        $('#header-level-' + levelId).attr('class', 'level_link ' + status);
+      }
+    });
+  });
+
```

Comment:
The small amount of benign duplication is OK, especially since we are rewriting as React.